### PR TITLE
Feature/minio s3

### PR DESCRIPTION
**Pull request by ryancraig**
All documentation, code, and tests required to allow Sleet to use Minio S3 object storage as feed persistence.

### CODE REVIEW

**Comment by emgarten** [comment]
This looks very complete, thank you @ryancraig 🥇 

I'm going to try this out

**Comment by ryancraig** [comment]
I did find one problem with it while I constructing a complete getting started like you did for the Azure feed. The public read-only policy is not correctly set on the index.json files. When I set this policy after the fact a `dotnet restore` works correctly. Anyway, I would appreciate your help in this matter. The AmazonS3FileSystem code has the necessary code to create the bucket, set it as public, and then set each file as public read only.

In conclusion, this issue should be resolved before the merge. I also need to push the integration test that tests this. Once this issue is resolved, I believe everything else is working well. I walked through the entire lifecycle of a NuGet package. 

I apologize that I was too quick to PR.

**Comment by ryancraig** [comment]
OK, after some more comprehensive manual testing, here are my results.
The changes I submitted in this PR work as I expect Sleet to work. Before `sleet init` you MUST create a bucket and set the bucket policy to public-readonly (out-of-band) manually or using some automated provisioning tool (Terraform, Ansible, your own...).

If the bucket does not exist, an error notifies you that the bucket does not exist and that you should create it before trying to initialize the Sleet feed.

If you create the bucket but do not set the bucket policy to public-readonly you get an error when you do a `dotnet restore`. You will get an exception about the index.json not containing valid json.

NOTE: I used Minio Client (mc), during my testing effort to create the bucket and set the bucket policy. `download` is minio short hand for public-readonly. You could also pass an AWS Policy json file.
```
./mc mb local/my-feed-bucket
./mc policy set download local/my-feed-bucket
```

In my opinion this behavior is correct. I personally would not want Sleet to create my Amazon buckets as I employ Terraform and Ansible as infrastructure provisioners and for automated configuration.

**Comment by ryancraig** [comment]
Well...I have have Sleet working with MinIO as a feed store. I am going to close this PR and open another once I'm very sure everything is working. It's just going to be easier to merge my changes in with the changes from the last merged PR to add the optional server side encryption config.

**Comment by emgarten** [comment]
@ryancraig you're going to make another PR that updates S3 instead of a new filesystem?

**Comment by ryancraig** [comment]
> @ryancraig you're going to make another PR that updates S3 instead of a new filesystem?
Yeah, I definitely don't want to push MinIO as a new filesystem into this project because it will make everyone suffer. As the S3 implementation needs to change, more than likely the MinIO implementation would too.
For my personal use, I am going down the same implementation path as you saw in my PR but I will refactor it in the near future. I just need it to work now.
I will notify you when my current implementation is complete as it'll show you what changes are required to make Sleet work with MinIO Server.  I'll keep my fork up-to-date with your project and work to modify the S3 implementation to work for all S3 API compliant stores including MinIO. As soon as I have a quality solution I'll PR or maybe someone else will beat me to it; )

**Comment by emgarten** [comment]
Thanks @ryancraig, keep me posted

**Comment by ryancraig** [comment]
@emgarten https://github.com/ryancraig/Sleet/tree/feature/minio-s3-feed
Here's the branch in my fork that includes the `minio` feed type. As I stated previously, I will use this for my needs now but I will actively work to modify the S3 feed type to work with Minio Server.
